Remove global group name tracking for ProcessGroupNCCL

pietern opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25905 Remove global group name tracking for ProcessGroupNCCL**

Now that we can detect and recover from failures in NCCL we should
allow processes that are started at different times (and perhaps have
had previous NCCL process group instances), to eventually be part of
the same process group. Keeping track of group names in global
variables prevents that, because the processes will be out of sync.

This commit removes the global group name maps and defers
responsibility of isolating access to the same store from multiple
process groups to the store itself. Users can use `c10d::PrefixStore`
to derive new store instances whose keyspace is scoped to some
prefix. Functionally, this is identical to keeping a global map and
using a group name, but also gives more flexibility to the front-end
API to reset state and have processes that have started at different
times to join the same process group.

Differential Revision: [D17281416](https://our.internmc.facebook.com/intern/diff/D17281416/)

pietern committed:
Remove global group name tracking for ProcessGroupNCCL

Now that we can detect and recover from failures in NCCL we should
allow processes that are started at different times (and perhaps have
had previous NCCL process group instances), to eventually be part of
the same process group. Keeping track of group names in global
variables prevents that, because the processes will be out of sync.

This commit removes the global group name maps and defers
responsibility of isolating access to the same store from multiple
process groups to the store itself. Users can use `c10d::PrefixStore`
to derive new store instances whose keyspace is scoped to some
prefix. Functionally, this is identical to keeping a global map and
using a group name, but also gives more flexibility to the front-end
API to reset state and have processes that have started at different
times to join the same process group.

Differential Revision: [D17281416](https://our.internmc.facebook.com/intern/diff/D17281416/)

pietern committed:
Update on "Remove global group name tracking for ProcessGroupNCCL"


Now that we can detect and recover from failures in NCCL we should
allow processes that are started at different times (and perhaps have
had previous NCCL process group instances), to eventually be part of
the same process group. Keeping track of group names in global
variables prevents that, because the processes will be out of sync.

This commit removes the global group name maps and defers
responsibility of isolating access to the same store from multiple
process groups to the store itself. Users can use `c10d::PrefixStore`
to derive new store instances whose keyspace is scoped to some
prefix. Functionally, this is identical to keeping a global map and
using a group name, but also gives more flexibility to the front-end
API to reset state and have processes that have started at different
times to join the same process group.

Differential Revision: [D17281416](https://our.internmc.facebook.com/intern/diff/D17281416/)

pietern commented:
The clang-tidy check is failing due to lines not changed in this PR.

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -257,11 +268,13 @@ class ProcessGroupNCCL : public ProcessGroup {
 
-  // The process group name
-  std::string groupName_;
+  // The number of NCCL communicators that have been create during
```
create -> created.

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -120,28 +120,39 @@ class ProcessGroupNCCL : public ProcessGroup {
+  // This constructor includes the deprecated `groupName` argument.
+  // If you have existing code that uses the `groupName`, you can replace
+  // it by specifying a `c10d::PrefixStore(groupName, store)` for store.
+  C10_DEPRECATED ProcessGroupNCCL(
```
Where is this constructor used? We are not exposing this to Python anyway, why do we need to keep a constructor with `groupName`?

pietern replied:
This is to not break downstream C++ users. I think torchcraft uses this directly.

pietern committed:
Update on "Remove global group name tracking for ProcessGroupNCCL"


Now that we can detect and recover from failures in NCCL we should
allow processes that are started at different times (and perhaps have
had previous NCCL process group instances), to eventually be part of
the same process group. Keeping track of group names in global
variables prevents that, because the processes will be out of sync.

This commit removes the global group name maps and defers
responsibility of isolating access to the same store from multiple
process groups to the store itself. Users can use `c10d::PrefixStore`
to derive new store instances whose keyspace is scoped to some
prefix. Functionally, this is identical to keeping a global map and
using a group name, but also gives more flexibility to the front-end
API to reset state and have processes that have started at different
times to join the same process group.

Differential Revision: [D17281416](https://our.internmc.facebook.com/intern/diff/D17281416/)
